lib/services/s3.js

```diff
@@ -479,6 +499,49 @@ AWS.util.update(AWS.S3.prototype, {
     }
   },
 
+  /**
+   * @api private
+   */
+  validateMd5TrailingChecksum: function validateMd5TrailingChecksum(resp) {
+    var response = resp.httpResponse;
+    var buffer = response.body;
+    var dataLength = buffer.length - 16;
+    var hash = buffer.slice(dataLength);
+    var data = buffer.slice(0, dataLength);
+    if (AWS.util.crypto.hash('md5', data, 'base64') === hash.toString('base64')) {
+      response.body = data;
+      if (response.headers['Content-Length']) response.headers['Content-Length'] = response.headers['Content-Length'] - 16;
+    } else {
+      throw AWS.util.error(
+        new Error('Response fails integrity check.'),
+        { code: 'NetworkingError' }
+      )
+    }
+  },
+
+  /**
+   * @api private
+   */
+  validateTrailingChecksum: function validateTrailingChecksum(resp) {
+    var response = resp.httpResponse;
+    if (
+      response.headers &&
+      response.headers['x-amz-transfer-encoding'] &&
+      response.headers['x-amz-transfer-encoding'].indexOf('append-') === 0
+    ) {
+      var hashAlgorithm = response.headers['x-amz-transfer-encoding'].substring('append-'.length);
```

Comment:
I think we also want the fall-back of checking the request header to see which algorithm we used. This will be more important in the browser where x-amz-transfer-encoding may not be exposed. Either that or we document which headers have to be exposed in the browser, which I'm also ok with.

Reply:
I will add a note to add the exposed header to documentation along with default value in config. Since the feature is defaulted to disabled on the browser, the users who want to enable the checksum they should be able to see documentation.